WordPress/Classes/Networking/DomainsRemote.swift

```diff
@@ -0,0 +1,155 @@
+    if let type = domainJson["type"] as? String
+        where type == "redirect" {
+        return .SiteRedirect
+    }
+
+    if let wpComDomain = domainJson["wpcom_domain"] as? Bool
+        where wpComDomain == true {
+        return .WPCom
+    }
+
+    if let hasRegistration = domainJson["has_registration"] as? Bool
+        where hasRegistration == true {
+        return .Registered
+    }
+
+    return .Mapped
+}
+
+struct DomainsService {
+    let remote: DomainsRemote
+    let blog: Blog
+    let siteID: Int
+
+    private let context = ContextManager.sharedInstance().mainContext
+
+    init(blog: Blog) {
+        precondition(blog.dotComID != nil)
+
+        remote = DomainsRemote(api: blog.restApi())
+        self.blog = blog
+        siteID = Int(blog.dotComID!)
+    }
+
+    func refreshBlogDomains(completion: (Bool) -> Void) {
+        remote.getDomainsForSite(siteID, success: { domains in
+            self.mergeDomains(domains)
+            }, failure: { error in
+                completion(false)
+        })
+    }
+
+    private func mergeDomains(domains: [Domain]) {
+        let remoteDomains = domains
+        let localDomains = blogDomains()
+
+        let remoteDomainNames = Set(remoteDomains.map({ $0.domain }))
+        let localDomainNames = Set(localDomains.map({ $0.domain }))
+
+        let removedDomainNames = localDomainNames.subtract(remoteDomainNames)
+        removeDomains(removedDomainNames)
+
+        // Let's try to only update objects that have changed
+        let remoteChanges = remoteDomains.filter {
+            return !localDomains.contains($0)
+        }
+
+        for remoteDomain in remoteChanges {
+            if let existingDomain = managedDomainWithName(remoteDomain.domain) {
+                existingDomain.updateWith(remoteDomain, blog: blog)
+                DDLogSwift.logDebug("Updated domain \(existingDomain)")
+            } else {
+                createManagedDomain(remoteDomain)
+            }
+        }
+
+        ContextManager.sharedInstance().saveContext(context)
+    }
+
+    private func managedDomainWithName(name: String) -> ManagedDomain? {
+        let request = NSFetchRequest(entityName: Domain.entityName)
+        request.predicate = NSPredicate(format: "blog = %@ AND domain = %@", blog, name)
+        request.fetchLimit = 1
+        let results = (try? context.executeFetchRequest(request) as! [ManagedDomain]) ?? []
+        return results.first
+    }
+
+    private func createManagedDomain(domain: Domain) {
+        let managedDomain = NSEntityDescription.insertNewObjectForEntityForName(Domain.entityName, inManagedObjectContext: context) as! ManagedDomain
+        managedDomain.updateWith(domain, blog: blog)
+        DDLogSwift.logDebug("Created domain \(managedDomain)")
+    }
+
+    private func blogDomains() -> [Domain] {
+        let request = NSFetchRequest(entityName: Domain.entityName)
+        request.predicate = NSPredicate(format: "%K == %@", "blog", blog)
```

Comment:
Copy pasta :) I'll add it straight into the string here – although in previous projects I've preferred using a parameter and an `Entity.entityName` property to avoid the chance of typos.